### PR TITLE
Inductor cpp wrapper: support None as output

### PR DESCRIPTION
**Pull request by chunyuan-w**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #90755
* #90754
* #89744
* #89743
* #89742
* #88667
* #88666
* #88561
* __->__ #88560

Map `None` to `at::Tensor()` in the cpp wrapper

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

### CODE REVIEW

File: torch/_inductor/codegen/wrapper.py [hunk 1]
```diff
@@ -468,8 +474,7 @@ def generate(self):
                 else:
                     self.wrapper_call.writeline(line)
 
-            output_refs = [x.codegen_reference() for x in V.graph.graph_outputs]
-            self.generate_return(output_refs)
+            self.generate_return()
```
Comment (jgong5) [inline torch/_inductor/codegen/wrapper.py:477]: Would it be simpler to add a `get_output_refs()` to be overridden by subclasses? The result can be cached with `lru_cache` for good perf. Then, you don't have to change `self.generate_return`.

File: torch/_inductor/codegen/wrapper.py [hunk 2]
```diff
@@ -558,6 +563,11 @@ def __init__(self):
         self._call_func_id = next(CppWrapperCodeGen.call_func_id)
         super().__init__()
 
+    def set_output_refs(self):
+        self.output_refs = [
+            x.cpp_wrapper_codegen_reference() for x in V.graph.graph_outputs
```
Comment (jgong5) [inline torch/_inductor/codegen/wrapper.py:568]: Maybe we can check the existence of `cpp_wrapper_codegen_reference` and call it or call `codegen_reference` instead here? With this, you don't have to add `cpp_wrapper_codegen_reference` to all the related IR nodes.

Comment (chunyuan-w) [reply torch/_inductor/codegen/wrapper.py]: Fixed as suggested.

File: torch/_inductor/codegen/wrapper.py [hunk 3]
```diff
@@ -313,8 +312,9 @@ def __init__(self):
             self.write_get_cuda_stream
         )
 
-    def set_output_refs(self):
-        self.output_refs = [x.codegen_reference() for x in V.graph.graph_outputs]
+    @functools.lru_cache(None)
```
Comment (jgong5) [inline torch/_inductor/codegen/wrapper.py:315]: `@cache_on_self`?

File: torch/_inductor/codegen/wrapper.py [hunk 4]
```diff
@@ -558,6 +562,20 @@ def __init__(self):
         self._call_func_id = next(CppWrapperCodeGen.call_func_id)
         super().__init__()
 
+    @functools.lru_cache(None)
```
Comment (jgong5) [inline torch/_inductor/codegen/wrapper.py:565]: ditto

**Review by jgong5 — CHANGES_REQUESTED** [review]

File: torch/_inductor/codegen/wrapper.py (hunk 3 above)
Comment (chunyuan-w) [reply torch/_inductor/codegen/wrapper.py]: Changed to `@cache_on_self`

File: torch/_inductor/codegen/wrapper.py (hunk 4 above)
Comment (chunyuan-w) [reply torch/_inductor/codegen/wrapper.py]: Changed to `@cache_on_self`

File: torch/_inductor/codegen/wrapper.py (hunk 1 above)
Comment (chunyuan-w) [reply torch/_inductor/codegen/wrapper.py]: Fixed as suggested.

**Review by jgong5 — APPROVED** [review]

**Review by jansel — APPROVED** [review]

**Review by desertfire — APPROVED** [review]

**Pull request closed without merging**